packages/lit-html/src/directive-helpers.ts

```diff
@@ -42,17 +48,35 @@ export const TemplateResultType = {
 export type TemplateResultType =
   (typeof TemplateResultType)[keyof typeof TemplateResultType];
 
+type isTemplateResultType = {
+  (val: unknown): val is TemplateResult | CompiledTemplateResult;
+  <T extends TemplateResultType>(
+    val: unknown,
+    type: T
+  ): val is TemplateResult<T>;
+};
+
 /**
- * Tests if a value is a TemplateResult.
+ * Tests if a value is a TemplateResult or a CompiledTemplateResult.
  */
-export const isTemplateResult = (
+export const isTemplateResult = ((
   value: unknown,
-  type?: TemplateResultType
-): value is TemplateResult =>
+  type: TemplateResultType | undefined
+) =>
   type === undefined
     ? // This property needs to remain unminified.
       (value as TemplateResult)?.['_$litType$'] !== undefined
-    : (value as TemplateResult)?.['_$litType$'] === type;
+    : (value as TemplateResult)?.['_$litType$'] ===
+      type) as isTemplateResultType;
```

Comment:
Is this `as isTemplateResultType` cast is in the wrong place? It reads to me like it's casting a boolean to a function call interface, but with a body-less arrow function I'm not sure of the precedence here.

To make it more clear I would type the `isTemplateResult` variable, instead of a cast, like:

```ts
export const isTemplateResult: IsTemplateResult = ((
  value: unknown,
  type: TemplateResultType | undefined
) =>
```

Reply:
Went through this with @rictic and unfortunately the cast is required. If I move the type as you've suggested it fails with: `Type '(value: unknown, type: TemplateResultType | undefined) => boolean' is not assignable to type 'IsTemplateResult'.`

Reply:
https://github.com/lit/lit/pull/3993#discussion_r1255004607

See the snippet here that moves the type next to the specifier. You need the 2nd param to be optional and also one of the type predicate return annotation to satisfy TS on the implementation code.

Reply:
Oh - I am so sorry. It did not click that you had removed the assertion!